### PR TITLE
Run Orion in the background

### LINKED ISSUE #6989: Add `prefect server start --detach` and `prefect server stop`
### First check

- [X] I added a descriptive title to this issue.
- [X] I used the GitHub search to find a similar request and didn't find it.
- [X] I searched the Prefect documentation for this feature.

### Prefect Version

2.x

### Describe the current behavior

The `prefect server start` command blocks until a keyboard interrupt is sent, then it is stopped.

### Describe the proposed behavior

Including a `--detach` command will allow the server to be run in the background. Then `prefect server stop` can be used to stop it.

### Example Use

```
prefect server start --detach
# Do some work with server...

prefect server stop
```

This would be useful in our compatibility tests e.g. #6988 

### Additional context

To facilitate this, we'll need to store the PID in `~/.prefect/` so during a start. We may want to do this in the server itself so that `prefect server stop` can interrupt servers that have been started without `--detach`.

We may want to take the `prefect server start` command and change it to be synchronous. It doesn't need to run multiple services in parallel anymore and it'd be easier to maintain that way. Our `open_process` implementation doesn't have a clear path to backgrounding a process.

### PR DESCRIPTION
**Pull request by hallenmaia**
### Summary

Adds `prefect orion start --detach` and `prefect orion stop` commands.

Closes #6989 and #6821

#### `prefect orion start --detach`:

Run Orion in the background. It can also be called using the shortcut `-d`

```shell
prefect orion start -d
Orion running in background.
Check out the dashboard at http://127.0.0.1:4200
```

A PID file is written to the path `~/.prefect/orion.pid` (depending on how `PREFECT_HOME` is set) so that it can be stopped.

An error is displayed when this file already exists and the `start` command is used:

```shell
prefect orion start
There is already an Orion process running in background.
Stop it with command `prefect orion stop`
```

#### `prefect orion stop`:

Checks if the file `~/.prefect/orion.pid` (depending on how `PREFECT_HOME` is configured) exists and terminates the background Orion process.

```shell
prefect orion stop    
Stopping Orion...
Orion stopped!
```

Displays an error if this file does not exist or is invalid. Example:

```shell
prefect orion stop
Orion is not running in the background.
```

### Checklist

- [x] This pull request references any related issue by including "closes `<link to issue>`"
	- If no issue exists and your change is not a small fix, please [create an issue](https://github.com/PrefectHQ/prefect/issues/new/choose) first.
- [x] This pull request includes tests or only affects documentation.
- [ ] This pull request includes a label categorizing the change e.g. `fix`, `feature`, `enhancement`

### CODE REVIEW

File: src/prefect/cli/orion.py [hunk 1]
```diff
@@ -114,39 +126,76 @@ async def start(
...
+        "--factory",
+        "prefect.orion.api.server:create_app",
+        "--host",
+        str(host),
+        "--port",
+        str(port),
+        "--access-log",
+    ]
+
+    pid_file = f"{PREFECT_HOME.value()}/orion.pid"
```
Comment (madkinsz) [inline src/prefect/cli/orion.py:142]: Need to use the `/` operator for windows support

```suggestion
    pid_file = str(PREFECT_HOME.value() / "orion.pid")
```

File: src/prefect/cli/orion.py [hunk 2]
```diff
@@ -114,39 +126,76 @@ async def start(
...
+            kill_on_interrupt(orion_process_id, "Orion", app.console.print)
 
-        kill_on_interrupt(orion_process_id, "Orion", app.console.print)
+            app.console.print("Orion stopped!")
 
-    app.console.print("Orion stopped!")
+
+@orion_app.command()
+async def stop():
+    pid_file = f"{PREFECT_HOME.value()}/orion.pid"
```
Comment (madkinsz) [inline src/prefect/cli/orion.py:186]: See above comment

File: src/prefect/cli/orion.py [hunk 3]
```diff
@@ -114,39 +126,76 @@ async def start(
...
+
+    if os.path.exists(pid_file):
+        try:
+            app.console.print("Stopping Orion...")
+            stop_process(pid_file)
+            exit_with_success("Orion stopped!")
+        except ValueError as e:
+            exit_with_error(e)
+        except ProcessLookupError:
+            exit_with_error(f"Orion process not found!")
```
Comment (madkinsz) [inline src/prefect/cli/orion.py:196]: Let's remove the PID file so they are allowed to run `prefect orion start` after?

File: src/prefect/cli/orion.py [hunk 4]
```diff
@@ -114,39 +126,76 @@ async def start(
...
+async def stop():
+    pid_file = f"{PREFECT_HOME.value()}/orion.pid"
+
+    if os.path.exists(pid_file):
+        try:
+            app.console.print("Stopping Orion...")
+            stop_process(pid_file)
+            exit_with_success("Orion stopped!")
+        except ValueError as e:
+            exit_with_error(e)
```
Comment (madkinsz) [inline src/prefect/cli/orion.py:194]: What kind of value errors are we catching here? :)

File: src/prefect/cli/orion.py (hunk 3 above)
Comment (madkinsz) [reply src/prefect/cli/orion.py:196]: Oh I see this happens in `stop_process` — it'd be good to include a note about that here

File: src/prefect/cli/orion.py (hunk 4 above)
Comment (madkinsz) [reply src/prefect/cli/orion.py:194]: I see now in `stop_process` as well this indicates an unparsable PID file. I'd note that here as well.

**Comment by hallenmaia** [comment]
@madkinsz I improved the error messages. The following tasks are pending:

- [ ] Create `test_orion.py` to improve coverage.
- [ ] Make the `start` command synchronous.

If there's anything else let me know.

**Comment by hallenmaia** [comment]
CTRL+C doesn't work on Uvicorn for Windows :(
Perhaps the `start` command has to remain asynchronous.

**Comment by hallenmaia** [comment]
Tests run on Windows with version:

```
Version:             0+untagged.12757.ga5c67cb
API version:         0.8.1
Python version:      3.10.7
Git commit:          a5c67cb0
Built:               Wed, Oct 5, 2022 10:36 AM
OS/Arch:             win32/AMD64
Profile:             default
Server type:         ephemeral
Server:
  Database:          sqlite
  SQLite version:    3.37.2
```

The `prefect orion start -d` and `prefect orion stop` commands worked perfectly. The error mentioned in #6821 does not occur with `--detach`.

**Comment by hallenmaia** [comment]
For now it will not be possible to change the `start` command to asynchronous because there is a problem with propagating exceptions with `uvicorn`.

https://github.com/encode/uvicorn/issues/1579

On POSIX systems it can be synchronous but on Windows systems CTRL+C is not sent to the `uvicorn` process. In the comment https://github.com/encode/uvicorn/issues/1579#issuecomment-1198051516 it is recommended to leave the process asynchronous to work around this issue.

**Comment by madkinsz** [comment]
Thanks so much for digging into this!

To clarify, the broken case here is `prefect orion start` without `--detach` cannot be stopped by Ctrl-C on Windows? All other cases are working?

I thought that the signal handling was fixed by @bunchesofdonald in #6672 but perhaps that doesn't work with the synchronous process implementation?

Do you see a clear path to retaining the async process implementation while returning before the process exists?

As an alternative, we could retain the async implementation for the blocking case and only use the sync implementation for `--detach`?

**Comment by hallenmaia** [comment]
> To clarify, the broken case here is prefect orion start without --detach cannot be stopped by Ctrl-C on Windows? All other cases are working?

Yup. CTRL+C works in all cases, but on Windows only if the function is asynchronous.

> I thought that the signal handling was fixed by @bunchesofdonald in #6672 but perhaps that doesn't work with the synchronous process implementation?

That's right. It was fixed in asynchronous mode. In synchronous mode `subprocess` sends all signals from the parent process to the children by default. It is not necessary to use `signals` for this.

But there is a problem in uvicorn's implementation of asyncio that is affecting this behavior on Windows.

I did the following test: I ran the uvicorn command in the Windows terminal inside the Prefect virtual environment and tried to close it using CTRL+C several times. As it didn't close I had to use the task manager for that.

> Do you see a clear path to retaining the async process implementation while returning before the process exists?

I did not understand that question.

> As an alternative, we could retain the async implementation for the blocking case and only use the sync implementation for --detach?

Yup. I believe this is the best solution until the `uvicorn` fix comes out.

**Comment by madkinsz** [comment]
:D by 

> Do you see a clear path to retaining the async process implementation while returning before the process exists?

I mean, can we use the previous async `open_process` implementation for  both `--detach` and the blocking start? I tried briefly but it would not return until the subprocess exited. If it's possible, it'd be nice not to split it into two implementations. Otherwise, I guess we'll split it into a detach implementation (sync) and a blocking implementation (async).

**Comment by hallenmaia** [comment]
Now I get it! :)

It is possible by removing the context manager from `open_process`. When you use the `anyio.open_process` function with context managers via the `with` statement it waits for the sub-process to finish before continuing.

I agree not to split it into two implementations and follow the [AnyIO documentation](https://anyio.readthedocs.io/en/stable/subprocesses.html):

> To run an external command with one call, use run_process().

> When you have more complex requirements for your interaction with subprocesses, you can launch one with open_process().

I'll go back to PR for draft and test it that way.

**Comment by hallenmaia** [comment]
Now closes #6821

**Comment by hallenmaia** [comment]
I merged the functions and now they have the following purposes:

`run_process`: To run an external command with a call.

`open_process`: When you have more complex requirements for your interaction with subprocesses.

### Windows

The `start` command works in both attached and detached mode. Issue #6821 been fixed.

For detached mode, there is an issue with the `ProactorEventLoop` implementation reported [here](https://github.com/python/cpython/issues/92841) which causes an exception to be thrown after the command goes into the background but this is not affecting the functioning of `uvicorn` (and other commands).

We have a few ways to go:

- [ ] Do not allow Orion to go in the background on Windows;
- [ ] Create two functions: `async_open_process` and `open_process` (synchronous) since in synchronous mode this exception does not occur.

**Comment by madkinsz** [comment]
Sweet! This is coming along nicely. 

Regarding Windows, I'm fine with users needing Python 3.10+ to not see the error. Especially since it does not actually impact the server.

I'm a little worried about leaving the async resources unclosed on cancellation. This was a big pain for me to fix previously and I do not want regressions. I think we also need async support for the process killing functionality (that is using `psutil`). Otherwise, we'll block the event loop / need to run in a worker thread. I'm going to do some investigation into these.

**Comment by hallenmaia** [comment]
Thank you very much. 😊

Sometimes I think it makes no sense to have a process detached asynchronously 🧐. That would avoid *some problems*.

**Pull request closed without merging**